plusdeck/dbus/interface.py

```diff
@@ -0,0 +1,170 @@
+import asyncio
+from typing import Optional, Self
+
+from sdbus import (  # pyright: ignore [reportMissingModuleSource]
+    dbus_method_async,
+    dbus_property_async,
+    dbus_signal_async,
+    DbusInterfaceCommonAsync,
+)
+
+from plusdeck.client import Client, create_connection, Receiver, State
+from plusdeck.config import Config
+
+DBUS_NAME = "org.jfhbrook.plusdeck"
+
+
+async def load_client(config_file: str) -> Client:
+    config: Config = Config.from_file(config_file)
+
+    client = await create_connection(config.port)
+
+    return client
+
+
+class DbusInterface(  # type: ignore
+    DbusInterfaceCommonAsync, interface_name=DBUS_NAME  # type: ignore
+):
+
+    def __init__(self: Self, config_file: str, client: Client) -> None:
+        super().__init__()
+        self._config_file = config_file
+        self._client: Client = client
+        self._client_lock: asyncio.Lock = asyncio.Lock()
+        self._rcv: Optional[Receiver] = None
+        self.subscribe()
+
+    @dbus_property_async("s")
+    def config_file(self: Self) -> str:
+        return self._config_file
+
+    def subscribe(self: Self) -> None:
+        self._subscription = asyncio.create_task(self.subscription())
+
+    async def subscription(self: Self) -> None:
+        if not self._rcv:
+            self._rcv: Optional[Receiver] = await self.client.subscribe()
+
+        while True:
+            if not self._rcv:
+                break
+            try:
+                state: State = await self._rcv.get_state()
+                self.state.emit(state.name)  # type: ignore
+            except TimeoutError:
+                pass
+
+    async def close(self: Self) -> None:
+        async with self._client_lock:
+            await self._close()
+
+    async def _close(self: Self, reloading=False) -> None:
+        client = self.client
+
+        if not reloading:
+            await client.unsubscribe()
+
+        self._rcv = None
+        await self._subscription
+        client.close()
+        await client.closed
+
+    async def _closed(self: Self) -> None:
+        # Shenanigans to allow for a refresh that replaces the client
+        client = self.client
+        while not client.closed.done():
+            await client.closed
+            client = self.client
+
+    @property
+    def closed(self: Self) -> asyncio.Future:
+        return asyncio.ensure_future(self._closed())
+
+    @dbus_method_async()
```

Comment:
This, plus `_closed`, need tests. This logic is really hairy!

Alternately, is there a pattern for dbus to reload by killing the service?

Reply:
I believe `systemctl restart some-dbus-service` will work. That is probably the correct way to manage a reload, especially considering that downtime is not a major concern.

Reply:
My proposal:

1. Confirm that the client as-is mostly works
2. Remove the reload method and simplify close behavior
3. Add a `--no-restart` flag to the config commands
4. if `--restart`, run `systemctl restart plusdeck.service`

Reply:
Leaving restart to the user, since they may have a configuration that doesn't use systemd.